sqsdomain/cosmwasmpool/orderbook.go

```diff
@@ -39,7 +122,65 @@ type OrderbookTickLiquidity struct {
 	AskLiquidity osmomath.BigDec `json:"ask_liquidity"`
 }
 
+// Returns the related liquidity for a given direction on the current tick
+func (tl *OrderbookTickLiquidity) ByDirection(direction OrderbookDirection) osmomath.BigDec {
+	if direction == ASK {
+		return tl.AskLiquidity
+	} else { // BID
+		return tl.BidLiquidity
+	}
+}
+
+// Determines how much of a given amount can be filled by the current tick state (independent for each direction)
+func (tl *OrderbookTickLiquidity) GetFillableAmount(input osmomath.BigDec, direction OrderbookDirection) osmomath.BigDec {
+	tickLiquidity := tl.ByDirection(direction)
+	if input.LT(tickLiquidity) {
+		return input
+	}
+	return tickLiquidity
+}
+
 type OrderbookTick struct {
 	TickId        int64                  `json:"tick_id"`
 	TickLiquidity OrderbookTickLiquidity `json:"tick_liquidity"`
 }
+
+// CalcAmountInToExhaustOrderbookLiquidity calculates the amount of token in needed to exhaust all liquidity in the orderbook.
+// - orderDirection is the direction of the order to be placed and liquidity to be exhausted will be in the opposite direction.
+// - startingIndex is the index of the tick to start the calculation from.
+// - ticks is the list of ticks in the orderbook, assumed to be ordered by tick id in ascending order.
+func CalcAmountInToExhaustOrderbookLiquidity(directionIn OrderbookDirection, startingIndex int, ticks []OrderbookTick) (osmomath.BigDec, error) {
```

Comment:
Is there a node-side PR with this change that I could review?

Has this already been tested on the node?

Reply:
yes